content/posts/0.27-release.md

```diff
@@ -0,0 +1,251 @@
+an error message that we are planning to throw away.
+
+The second improvement we made was to optimize away temporary bound methods.
+Python code like
+```py
+a.f()
+```
+means first look up `a.f` and then call the result, which is translated into
+bytecode like:
+```
+LOAD a
+LOAD_ATTR f
+CALL
+```
+The function `a.f` receives `a` as the first `self` argument, so it needs to be
+a special bound method object that knows about the `self` argument. However, we
+only use this object once to call it and then we are going to throw it away. We
+can avoid allocating and destroying an object when we call a method by calling
+`type(a).f(a)`. The `LOAD_ATTR` opcode has a special argument that indicates
+that the next opcode is going to be `CALL` and in that case calls a method named
+`_PyObject_GetMethod` instead of the typical `PyObject_GetAttr` to perform the
+attribute lookup. We patched `_PyObject_GetMethod` to have special handling for
+`JsProxy` objects so that we could optimize away the temporary for JS objects too.
+
+### A potential future performance improvement
+
+In the case of lookup on a `JsProxy`, there are other bytecode patterns where we
+could optimize away temporaries. For instance, if `a` is a `JsProxy` for the
+JavaScript object `{ x: { y: 2 } }` then `a.x.y` creates a temporary `JsProxy`
+pointing to `{ y: 2 }` and then looks up `y` on the temporary and returns `2`.
+The bytecode looks like:
+```
+LOAD a
+LOAD_ATTR x
+LOAD_ATTR y
+```
+If we know that we were going to use the result of the first `LOAD_ATTR x`
+immediately, we could avoid creating the temporary. Unfortunately, unlike the
+method call example, native Python code doesn't create a temporary here so
+optimizing this would involve more invasive patches to the interpreter than
+removing the temporary bound methods.
+
+## What's next?
+
+Our next release should use Python 3.13. In addition, we are planning to make
+the following changes:
+
+### Package unvendoring
+
+We will unvendor the package recipes from the main Pyodide repository. The
+package index will gain a separate release process from the Pyodide runtime
+itself. This should have several significant benefits:
+
+* It will reduce CI usage due to rebuilding packages less often.
+* People who add a package will not have to wait as long until their package is
+  included in our index.
+
+### Upstream work in Emscripten and Python
+
+We are working on restoring Emscripten to a tier 3 supported target for Python.
+As a part of this work, we are upstreaming a large number of fixes to the
+Emscripten file system. For example, it now works to seek on `/dev/null` (it
+does nothing), symlink support is much better, it will work to `stat` a file
+descriptor which doesn't point to a named file, and many file system system
+calls have more posix-compliant error handling. Most of these changes have not
+made their way into Pyodide yet because we're still using Emscripten 3.1.68 for
+ABI compatibility.
```

Comment:
```suggestion
calls have more posix-compliant error handling. Most of these changes have not
made their way into Pyodide yet because we're still using Emscripten 3.1.58 for
ABI compatibility.
```

(probably a typo?)